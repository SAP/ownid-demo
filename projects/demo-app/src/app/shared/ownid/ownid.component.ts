import { ConsoleLogger } from '@services/console-logger.service';
import {
  Component,
  OnInit,
  ChangeDetectionStrategy,
  Input,
  ElementRef,
  Output,
  EventEmitter,
  OnDestroy
} from "@angular/core";
import WidgetComponent from '../../../../../../src/assets/ownid-web-ui-sdk/components/widget.component';
import { environment } from '../../../environments/environment';

@Component({
  selector: "ownid",
  template: "",
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class OwnidComponent implements OnInit, OnDestroy {
  @Input() type: string | null = null;

  @Input() data: unknown | null = null;

  @Output() onLogin = new EventEmitter();

  @Output() onRegister = new EventEmitter();

  @Output() onLink = new EventEmitter();

  @Output() onRecover = new EventEmitter();

  private ownidWidget: WidgetComponent | undefined;

  constructor(private elRef: ElementRef) {}

  async ngOnInit() {
    // @ts-ignore-next-line
    window.ownid!.init({
      statusInterval: 1000,
      URLPrefix: environment.ownidURLPrefix,
      logger: new ConsoleLogger(),
      logLevel: 'info',
    });

    if (this.type === 'link') {
      // @ts-ignore-next-line
      this.ownidWidget = await window.ownid!.renderLinkGigya({
        element: this.elRef.nativeElement,
        type: this.type,
        onLink: this.onLink.emit.bind(this.onLink),
      }, environment.gigyaApiKey);
    } else {
      // @ts-ignore-next-line
      this.ownidWidget = window.ownid!.render({
        element: this.elRef.nativeElement,
        type: this.type,
        data: this.data,
        onLogin: this.onLogin.emit.bind(this.onLogin),
        onRegister: this.onRegister.emit.bind(this.onRegister),
        onLink: this.onLink.emit.bind(this.onLink),
        onRecover: this.onRecover.emit.bind(this.onRecover),
      });
    }
  }

  ngOnDestroy() {
    this.ownidWidget?.destroy();
  }
}
