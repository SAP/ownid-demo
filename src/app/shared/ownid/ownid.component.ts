import {
  Component,
  OnInit,
  ChangeDetectionStrategy,
  Input,
  ElementRef,
  Output,
  EventEmitter,
} from '@angular/core';

@Component({
  selector: 'ownid',
  template: '',
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class OwnidComponent implements OnInit {
  @Input() type: string | null = null;

  @Output() onLogin = new EventEmitter();

  @Output() onRegister = new EventEmitter();

  constructor(private elRef: ElementRef) {}

  ngOnInit(): void {
    // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
    // @ts-ignore-next-line
    window.ownid!.init({
      element: this.elRef.nativeElement,
      type: this.type,
      onLogin: this.onLogin.emit.bind(this.onLogin),
      onRegister: this.onRegister.emit.bind(this.onRegister),
      statusInterval: 3000,
      getContextURL: "/netcore3/ownid",
      getStatusURL: "/netcore3/ownid/:context/status"
    });
  }
}
