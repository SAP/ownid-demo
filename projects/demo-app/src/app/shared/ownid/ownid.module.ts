import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { OwnidComponent } from './ownid.component';

@NgModule({
  declarations: [OwnidComponent],
  exports: [OwnidComponent],
  imports: [CommonModule],
})
export class OwnidModule {}
