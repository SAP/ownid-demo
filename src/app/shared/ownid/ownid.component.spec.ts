import { async, TestBed } from "@angular/core/testing";
import { ElementRef, NO_ERRORS_SCHEMA } from "@angular/core";
import { OwnidComponent } from "./ownid.component";
import WidgetComponent from "../../../assets/ownid-web-ui-sdk/components/widget.component";

describe("OwnidComponent", () => {
  describe("Snapshot", () => {
    beforeEach(async(() => {
      TestBed.configureTestingModule({
        declarations: [OwnidComponent],
        schemas: [NO_ERRORS_SCHEMA]
      }).compileComponents();
    }));

    test("check snapshot", () => {
      const fixture = TestBed.createComponent(OwnidComponent);

      expect(fixture).toMatchSnapshot();
    });
  });

  describe("ngOnInit", () => {
    it("should call window.ownid.init", () => {
      const el = { nativeElement: { el: true } } as ElementRef;
      // @ts-ignore-next-line
      window.ownid = {
        init: jest.fn(),
        render: jest.fn()
      };

      const sut = new OwnidComponent(el);

      sut.type = "login";

      sut.ngOnInit();
      // @ts-ignore-next-line
      expect(window.ownid.init).toBeCalled();
      // @ts-ignore-next-line
      expect(window.ownid.render).toBeCalled();
    });

    it("should call window.ownid!.init", () => {
      const el = { nativeElement: { el: true } } as ElementRef;
      // @ts-ignore-next-line
      window.ownid = {
        init: jest.fn(),
        render: jest.fn(),
        renderLinkGigya: jest.fn()
      };

      const sut = new OwnidComponent(el);

      sut.type = "link";

      sut.ngOnInit();
      // @ts-ignore-next-line
      expect(window.ownid!.renderLinkGigya).toBeCalled();
    });
  });

  describe("ngOnDestroy", () => {
    it("should call destroy", () => {
      const el = { nativeElement: { el: true } } as ElementRef;
      // @ts-ignore-next-line
      window.ownid = {
        init: jest.fn(),
        render: jest.fn()
      };

      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      const sut: any = new OwnidComponent(el);

      sut.type = "login";

      sut.ownidWidget = {} as WidgetComponent;
      sut.ownidWidget.destroy = jest.fn();

      sut.ngOnDestroy();

      expect(sut.ownidWidget.destroy).toBeCalled();
    });
  });
});
