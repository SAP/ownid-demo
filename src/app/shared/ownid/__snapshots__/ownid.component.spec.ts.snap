// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OwnidComponent Snapshot check snapshot 1`] = `
<ownid
  data="null"
  elRef={[Function ElementRef]}
  onLink={[Function EventEmitter_]}
  onLogin={[Function EventEmitter_]}
  onRecover={[Function EventEmitter_]}
  onRegister={[Function EventEmitter_]}
  partial="false"
  type="null"
>
  
</ownid>
`;
